scripts/peer-vnets.sh

```diff
@@ -0,0 +1,426 @@
+    az network private-dns zone wait \
+        --resource-group "${CLUSTER_NAME}" \
+        --name "${DNS_ZONE}" \
+        --created --timeout "${WAIT_TIMEOUT}" \
+        --only-show-errors --output none || error "Timeout waiting for private DNS zone"
+    print_step "1/4" "${DNS_ZONE} private DNS zone created in ${CLUSTER_NAME}"
+
+    # Link private DNS Zone to workload vnet with improved error handling
+    az network private-dns link vnet create \
+        --resource-group "${CLUSTER_NAME}" \
+        --zone-name "${DNS_ZONE}" \
+        --name "${CLUSTER_NAME}-to-mgmt" \
+        --virtual-network "${WORKLOAD_VNET_ID}" \
+        --registration-enabled false \
+        --only-show-errors --output none || error "Failed to create workload DNS link"
+    az network private-dns link vnet wait \
+        --resource-group "${CLUSTER_NAME}" \
+        --zone-name "${DNS_ZONE}" \
+        --name "${CLUSTER_NAME}-to-mgmt" \
+        --created --timeout "${WAIT_TIMEOUT}" \
+        --only-show-errors --output none || error "Timeout waiting for workload DNS link"
+    print_step "2/4" "workload cluster vnet ${CLUSTER_NAME}-vnet linked with private DNS zone"
+
+    # Link private DNS Zone to mgmt vnet with improved error handling
+    az network private-dns link vnet create \
+        --resource-group "${CLUSTER_NAME}" \
+        --zone-name "${DNS_ZONE}" \
+        --name "mgmt-to-${CLUSTER_NAME}" \
+        --virtual-network "${MGMT_VNET_ID}" \
+        --registration-enabled false \
+        --only-show-errors --output none || error "Failed to create management DNS link"
+    az network private-dns link vnet wait \
+        --resource-group "${CLUSTER_NAME}" \
+        --zone-name "${DNS_ZONE}" \
+        --name "mgmt-to-${CLUSTER_NAME}" \
+        --created --timeout "${WAIT_TIMEOUT}" \
+        --only-show-errors --output none || error "Timeout waiting for management DNS link"
+    print_step "3/4" "management cluster vnet ${AKS_MGMT_VNET_NAME} linked with private DNS zone"
+
+    # Create private DNS zone record with improved error handling
+    az network private-dns record-set a add-record \
+        --resource-group "${CLUSTER_NAME}" \
+        --zone-name "${DNS_ZONE}" \
+        --record-set-name "@" \
+        --ipv4-address "${AZURE_INTERNAL_LB_PRIVATE_IP}" \
+        --only-show-errors --output none || error "Failed to create DNS record"
+    print_step "4/4" "\"@\" private DNS zone record created to point ${DNS_ZONE} to ${AZURE_INTERNAL_LB_PRIVATE_IP}"
+    print_success "Private DNS zone creation completed successfully"
+}
+
+# New function that waits for NSG rules with prefix "NRMS-Rule-101" in the relevant resource groups,
+# then creates or modifies NRMS-Rule-101 to allow the specified ports.
+wait_and_fix_nsg_rules() {
+    local tcp_ports="443 5986 6443"
+    local udp_ports="53 123"
+    local timeout=3000     # seconds to wait per NSG for the appearance of an NRMS-Rule-101 rule
```

Comment:
Is this timeout a bit excessive? I know the rule takes a while to be populated but not sure if 50 minutes is too much. This is probably not a big issue though since I don't think we'll ever wait that long, and the rule always gets populated by some Azure job.

Reply:
It is definitely on the higher end. But in my testing, a 3 control plane - 2 worker node cluster takes about 30 mins to come up. And the NSG rules can take some more time. 
`3000` seemed like a good number.